controller/src/db/sqldb.py

```diff
@@ -0,0 +1,630 @@
+# Copyright 2023 Iguazio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+import re
+from typing import List, Optional
+
+from sqlalchemy import (
+    JSON,
+    Column,
+    ForeignKey,
+    Index,
+    Integer,
+    String,
+    Table,
+    UniqueConstraint,
+)
+from sqlalchemy.ext.mutable import MutableDict
+from sqlalchemy.orm import (
+    Mapped,
+    declarative_base,
+    declared_attr,
+    mapped_column,
+    relationship,
+)
+
+ID_LENGTH = 64
+TEXT_LENGTH = 1024
+
+# Create a base class for declarative class definitions
+Base = declarative_base()
+
+
+def make_label(table):
+    class Label(Base):
+        __tablename__ = f"{table}_labels"
+        __table_args__ = (
+            UniqueConstraint("name", "parent", name=f"_{table}_labels_uc"),
+            Index(f"idx_{table}_labels_name_value", "name", "value"),
+        )
+
+        id = Column(Integer, primary_key=True)
+        name = Column(String(255, None))  # in mysql collation="utf8_bin"
+        value = Column(String(255, collation=None))
+        parent = Column(Integer, ForeignKey(f"{table}.name"))
+
+    return Label
+
+
+def update_labels(obj, labels: dict):
+    old = {label.name: label for label in obj.labels}
+    obj.labels.clear()
+    for name, value in labels.items():
+        if name in old:
+            old[name].value = value
+            obj.labels.append(old[name])
+        else:
+            obj.labels.append(obj.Label(name=name, value=value, parent=obj.name))
+
+
+class BaseSchema(Base):
+    """
+    Base class for all tables.
+    We use this class to define common columns and methods for all tables.
+
+    :arg  id: unique identifier for each entry.
+    :arg  name: entry's name.
+    :arg  description: The entry's description.
```

Comment:
You have a double space after `arg` in the docstring of each class for some reason

```suggestion
    :arg id: unique identifier for each entry.
    :arg name: entry's name.
    :arg description: The entry's description.
```